src/main/scala/io/iohk/ethereum/db/dataSource/LevelDBDataSource.scala

```diff
@@ -44,17 +37,25 @@ class LevelDBDataSource(
     */
   override def update(namespace: Namespace, toRemove: Seq[Key], toUpsert: Seq[(Key, Value)]): DataSource = {
     val batch = db.createWriteBatch()
-    toRemove.foreach { key => batch.delete((namespace ++ key).toArray) }
-    toUpsert.foreach { item => batch.put((namespace ++ item._1).toArray, item._2.toArray) }
-    db.write(batch, new WriteOptions())
+
```

Comment:
Did you investigate merge operator from rocksdb ([merge-operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator))? I wonder if it could improve performance.

Reply:
i remember investigating it some day as it would greatly improve our reference counting but unfortunatly it is not part of java api yet https://github.com/facebook/rocksdb/issues/2282